File: Week_01/G20200343030400/LeetCode_22-sec_400.cpp
//22
//https://leetcode-cn.com/problems/generate-parentheses

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string s = "";
        s.append(n,'(');
        s.append(n,')');
        string s_final = "";
        for (int i = 0; i < n; ++i)
            s_final += "()";

        do {
            if (isWellFormed(s))
                ans.push_back(s);
        } while (next_permutation(s.begin(), s.end()) && s <= s_final);
        return ans;
    }

    bool isWellFormed(string& s){
        int count = 0;
        for (int i = 0; i < s.size(); ++i) {
            if (s[i] == '(')
                ++count;
            else
                --count;
            if (count < 0)
                return false;
        }
        return count == 0;
    }
};

File: Week_01/G20200343030400/LeetCode_22-sec_400.py
//22
//https://leetcode.com/problems/generate-parentheses/

class Solution:

    def nextPermutation(self, lst):
        n = len(lst)
        i = n-2 
        while i>=0 and lst[i+1] <= lst[i]: i -= 1 
        if i >= 0: 
            j = n-1 
            while j >= 0 and lst[j] <= lst[i]: j -= 1
            lst[i], lst[j] = lst[j], lst[i] 
        lst[i+1:n] = list(reversed(lst[i+1:n]))
        return lst

    def isValid(self, paratheses):
        count = 0
        for char in paratheses:
            if char == "(":
                count += 1
            elif char == ")":
                count -= 1
                if count < 0:
                    return False
        return True

    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        s = list("(" * n + ")" * n)
        s_final = list("()" * n)

        while operator.le(s, s_final):
            if self.isValid(s):
                ans.append("".join(s));
            s = self.nextPermutation(s)
               
        return ans        
